core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java

```diff
@@ -1809,6 +1816,13 @@ private boolean needNewSubQuery(
         return true;
       }
 
+      if (rel instanceof Project
+          && clauses.contains(Clause.ORDER_BY)
+          && dialect.getConformance().isSortByOrdinal()) {
+        // Cannot merge a Project that contains sort by ordinal under it.
+        return hasSortByOrdinal();
+      }
```

Comment:
The [last version we merged](https://github.com/apache/calcite/commit/0e8d6b093d26c291361bfd63ecb14828d1e33bc6) does not have this logic, could you help me to understand why do you need to add this logic now?

Reply:
We found test case LatticeTest#testSum is failed in slow test.
These code is to fix it.

The reason of test case failure is that the project will be merged when the sub-relnode is sort(change by [CALCITE-4901](https://issues.apache.org/jira/browse/CALCITE-4901)).

For example,
If the relNode is
```
JdbcProject(C=[$1])
  JdbcSort(sort0=[$1], dir0=[DESC], fetch=[1])
    JdbcAggregate(group=[{0}], C=[$SUM0($7)])
      JdbcTableScan(table=[[foodmart, sales_fact_1997]])
```
The result after merge project is
```
SELECT SUM(`sales_fact_1997`.`unit_sales`) AS `C`
FROM `foodmart`.`sales_fact_1997` AS `sales_fact_1997`
GROUP BY `sales_fact_1997`.`product_id`
ORDER BY 2 DESC
```
it's a incorrect sql, we want
```
SELECT `C` FROM (
SELECT `sales_fact_1997`.`product_id`, SUM(`sales_fact_1997`.`unit_sales`) AS `C`
FROM `foodmart`.`sales_fact_1997` AS `sales_fact_1997`
GROUP BY `sales_fact_1997`.`product_id`
ORDER BY 2 DESC) t
```
So, these code to find out whether there is sort by original under project. 
And don't merge project when there is sort by original under project.

Reply:
Since we agreed that the slow tests failure is not caused by this PR, we'd better handle this in CALCITE-5537.

Reply:
There are three test cases failure.
1. LatticeTest#testSum
2. LatticeSuggesterTest#testFoodMartAll
3. LatticeSuggesterTest#testFoodMartAllEvolve

Case 1 is caused by this PR. Maybe we could fix in this PR?
Case 2 & 3 is not caused by this PR. I think we could fix case 2 & 3 in [CALCITE-5537](https://issues.apache.org/jira/browse/CALCITE-5537).

BTW, these code just fix case 1.

Reply:
Thanks for the clarification, I agree with you. I'll review the PR in the coming days.